Utils/CSV.cs

```diff
@@ -0,0 +1,308 @@
+				{
+					var value = int.Parse(s);
+					if (value < minValue) throw new ArgumentException("Min value not reached: " + maxValue);
+					if (value > maxValue) throw new ArgumentException("Max value exceeded: " + maxValue);
+					return value;
+				}, allowEmptyValues){}
+			}
+
+			public class Bool : Type<bool>
+			{
+				public Bool(bool allowEmptyValues = false) : base(bool.Parse, allowEmptyValues) { }
+			}
+
+			public class Float : Type<float>
+			{
+				public Float() : base(float.Parse) { }
+
+				public Float(bool allowEmptyValues = false, float minValue = float.MinValue, float maxValue = float.MaxValue) : base(s =>
+				{
+					var value = float.Parse(s);
+					if (value < minValue) throw new ArgumentException("Min value not reached: " + maxValue);
+					if (value > maxValue) throw new ArgumentException("Max value exceeded: " + maxValue);
+					return value;
+				}, allowEmptyValues){ }
+			}
+
+			public class Enum<T> : Type<T> where T : struct, IConvertible
+			{
+				public Enum(bool ignoreCase = false, bool allowEmptyValues = false, T fallbackValue = default(T)) : base(s =>
+				{
+					if (!typeof(T).IsEnum) throw new ArgumentException("Type must be an Enum type.");
+
+					if (string.IsNullOrEmpty(s))
+					{
+						if (allowEmptyValues) return fallbackValue;
+
+						throw new ArgumentNullException();
+					}
+
+					return (T)Enum.Parse(typeof(T), s, ignoreCase);
+				}, allowEmptyValues){ }
+			}
+		}
+
+		public class Format
+		{
+			private List<string> orderedKeys;
+			private Dictionary<string, Type> format;
+			private Type defaultType = new Types.String(allowEmptyValues: true);
+
+			public Format(Dictionary<string, Type> format, Type defaultType = null)
+			{
+				this.format = format;
+
+				if (defaultType != null)
+				{
+					this.defaultType = defaultType;
+				}
+			}
+
+			internal string GetFieldName(int index)
+			{
+				return (orderedKeys[index]);
+			}
+
+			internal Type GetType(string fieldName)
+			{
+				return format[fieldName];
+			}
+
+			internal void ValidateHeader(string[] header, bool allowUnexpectedFields = false)
+			{
+				if (format == null || format.Keys.Count == 0) return;
+
+				var expectedFieldNames = new List<string>(format.Keys);
+
+				orderedKeys = new List<string>();
+
+				foreach (var textChunk in header)
+				{
+					var fieldName = textChunk.Trim();
+					var didFind = expectedFieldNames.Remove(fieldName);
+
+					if (!didFind)
+					{
+						// Field is known but wasn't remaining in the 'expected' - i.e. we've removed it before
+						if (format.ContainsKey(fieldName))
+						{
+							throw new Exception("Duplicate field: " + fieldName);
+						}
+						// Field isn't known
+						else
+						{
+							if (!allowUnexpectedFields)
+							{
+								throw new Exception("Unknown field: " + fieldName);
+							}
+							else
+							{
+								format.Add(fieldName.Trim(), defaultType);
```

Comment:
Is there a standard approach in DUCK in the way that default failover behaviour is communicated to it's user. For example I think that here we should log out a warning in the console, saying that this field name will default failover and print what that value would be.

Reply:
Sounds good, I'll add a warning